Fix FSYNC logic

ashwinstar opened the pull request:
PR fixes issue discussed on mailing list at:
https://groups.google.com/a/greenplum.org/forum/#!topic/gpdb-dev/PHKuQPNwWs0

High level:
- Correctly maintain pendingOpsTable in checkpoint process
- Prevent deadlock between a committing backend and checkpointer process
- Has validation tests accompanying the fix

Full (fork of) master pipeline was run against the change and all is GREEN.

The majority of the work on fixing this was done by @asimrp in co-ordination with @xinzweb. Thanks a lot to @hlinnaka for reporting the problem in depth and continued guidance along the way.

hlinnaka commented on src/backend/access/transam/xlog.c:
```diff
@@ -9044,6 +9044,11 @@ CreateCheckPoint(int flags)
 	{
 		do
 		{
+			/* This is to avoid deadlock when fsync request queue is full
```
Would be good to mention why we need this, while PostgreSQL doesn't. So that we know when we can remove this in the future, if it becomes unnecessary.

hlinnaka commented on src/backend/postmaster/bgwriter.c:
```diff
@@ -751,8 +735,7 @@ AbsorbFsyncRequests(void)
 
-	if (!am_bg_writer)
-		return;
+	Insist(!IsUnderPostmaster || AmStartupProcess() || AmCheckpointerProcess());
```
Insist produces hard-to-debug errors. I'd prefer the usual upstream pattern with if() elog. Something like:
```
  if (IsUnderPostmaster && !AmStartupProcess() && !AmCheckpointerProcess())
    elog(ERROR, "AbsorbFsyncRequests() called in process %d (type %d)", MyProcPid, MyAuxProcType);
```

ashwinstar commented:
On Fri, Mar 3, 2017 at 12:35 AM, Heikki Linnakangas <
notifications@github.com> wrote:

> *@hlinnaka* commented on this pull request.
>
> It's a bit difficult to see if we've missed something, because the GPDB
> code is a mix & match of PG 8.3 parts, and PG 9.2 parts. For example, in PG
> 9.2, when the checkpointer was separated from bgwriter, the responsibility
> for switching xlog files was also moved to the checkpointer, but in GPDB,
> it's still the bgwriter's responsibility. That's correct, AFAICS, but this
> kind of differences make it more difficult to verify.
>
> Also, I'd suggest renaming src/backend/postmaster/checkpoint.c, to
> checkpointer.c, like it is in PG 9.2. That'll also make it easier to do
> "git diff pgsql/REL9_2_STABLE".
>
Sure, I can do that. But essentially will come with lot many changes as all
related data-structures were renamed to Checkpointer as well like
`CheckpointShmemStruct` to `CheckpointerShmemStruct`. Seem fine to go ahead
to get them fixed, right ?

> There probably should be an AbsorbFsyncRequests() call in the main loop of
> the checkpointer, like there is in PG 9.2. Not a correctness issue, but
> would help to avoid filling up the queue. (That's a good example of the
> kind of things that's easy to miss, when the code is a mix & match of
> different upstream versions.)
>
Yes, we definitely need this. Personally saw many times after ICW drop
database waiting for long time for checkpoint to kick-in and clean off the
queue. To some extent missing it though till now helped us hit queue full
deadlocks :-)

ashwinstar commented:
@hlinnaka addressed your review comments. Please let me know if anything else.
The comment about checkpointer file change is needed agreed, planning to take that up as separate PR once this makes in. As this functionally fixes the broken thing and helps enable yet to be enabled CI suites awaiting for the fix.

ashwinstar commented:
@hlinnaka Made the change for file change to checkpointer and corresponding changes, tried to bring it as close as to REL9_2_STABLE. Please review.

hlinnaka commented on src/backend/storage/smgr/md.c:
```diff
@@ -41,7 +41,7 @@
  * Special values for the segno arg to RememberFsyncRequest.
  *
- * Note that CompactBgwriterRequestQueue assumes that it's OK to remove an
+ * Note that CompactcheckpointerRequestQueue assumes that it's OK to remove an
```
should be CompactCheckpointerRequestQueue, with capital C

hlinnaka commented on src/include/access/xlog_internal.h:
```diff
@@ -269,7 +269,7 @@ typedef struct RmgrData
 
 /*
- * Exported to support xlog switching from bgwriter
+ * Exported to support xlog switching from chckpointer
```
s/chckpointer/checkpointer/
